File: src/Segregate 0s and 1s/Code/solution.js
/**
 * @param {number[]} arr
 */
class Solution {
  segregate0and1(arr) {
    let left = 0;
    let right = arr.length - 1;

    // Continue until both pointers cross each other
    while (left < right) {
      // Move left pointer forward if current value is 0
      while (left < right && arr[left] === 0) {
        left++;
      }

      // Move right pointer backward if current value is 1
      while (left < right && arr[right] === 1) {
        right--;
      }

      // Swap misplaced elements
      if (left < right) {
        [arr[left], arr[right]] = [arr[right], arr[left]];
        left++;
        right--;
      }
    }
  }
}

File: src/Segregate 0s and 1s/Code/solution.py
class Solution:
    def segregate0and1(self, arr):
        left = 0
        right = len(arr) - 1

        # Continue until both pointers cross each other
        while left < right:

            # Move left pointer if current value is 0
            while left < right and arr[left] == 0:
                left += 1

            # Move right pointer if current value is 1
            while left < right and arr[right] == 1:
                right -= 1

            # Swap misplaced values
            if left < right:
                arr[left], arr[right] = arr[right], arr[left]
                left += 1
                right -= 1